lib/fluent/plugin/out_secondary_file.rb

```diff
@@ -0,0 +1,151 @@
+    config_param :compress, :enum, list: [:normal, :gz, :gzip], default: :normal
+
+    def configure(conf)
+      super
+
+      unless @as_secondary
+        raise Fluent::ConfigError, "This plugin can only be used in the <secondary> section"
+      end
+
+      configure_path!
+
+      test_path = generate_path(Time.now.strftime("%Y%m%d"))
+      unless Fluent::FileUtil.writable_p?(test_path)
+        raise Fluent::ConfigError, "out_secondary_file: `#{test_path}` is not writable"
+      end
+
+      @dir_perm = system_config.dir_permission || DIR_PERMISSION
+      @file_perm = system_config.file_permission || FILE_PERMISSION
+    end
+
+    def write(chunk)
+      id = generate_id(chunk)
+      path = generate_path(id)
+      FileUtils.mkdir_p File.dirname(path), mode: @dir_perm
+
+      case @compress
+      when :normal
+        File.open(path, "ab", @file_perm) {|f|
+          f.flock(File::LOCK_EX)
+          chunk.write_to(f)
+        }
+      when :gz, :gzip
+        File.open(path, "ab", @file_perm) {|f|
+          f.flock(File::LOCK_EX)
+          gz = Zlib::GzipWriter.new(f)
+          chunk.write_to(gz)
+          gz.close
+        }
+      end
+
+      path
+    end
+
+    private
+
+    def generate_id(chunk)
+      if chunk.metadata.empty?
+        chunk.chunk_id
+      else
+        extract_placeholders(@path, chunk.metadata)
+      end
+    end
+
+    def configure_path!
+      # Check @path includes ${tag} etc.
+      matched = @path.scan(/\${([\w.@-]+(\[\d+\])?)}/).flat_map(&:first) # to trim suffix [\d+]
+
+      if path_has_tags(matched) || path_has_time_format?
+        validate_path_is_comptible_with_primary_buffer!(matched)
+        @path_prefix = ""
+        @path_suffix = ".log"
+      else
+        if pos = @path.index('*')
+          @path_prefix = @path[0, pos]
+          @path_suffix = @path[pos+1..-1]
+        else
+          @path_prefix = @path + "."
+          @path_suffix = ".log"
+        end
+      end
+    end
+
+    def validate_path_is_comptible_with_primary_buffer!(matched)
+      raise "BUG: file path has imcompatible placeholder: Time" if !@chunk_key_time && path_has_time_format?
+      matched.each do |e|
+        case
+        when @chunk_key_tag && e =~ /tag(\[\d+\])?/
+        # ok
+        when !@chunk_key_tag && e =~ /tag(\[\d+\])?/
+          raise "BUG: file path has imcompatible placeholder: #{e}"
+        when @chunk_keys && @chunk_keys.include?(e)
+        # ok
+        else
+          raise "BUG: file path has imcompatible placeholder: #{e}"
+        end
+      end
+    end
+
+    def path_has_tags(matched)
+      !matched.empty?
+    end
+
+    def path_has_time_format?
+      @path != Time.now.strftime(@path)
+    end
+
+    def suffix
+      case @compress
+      when :normal
+        ""
+      when :gz, :gzip
+        ".gz"
+      end
+    end
+
+    def generate_path(id)
+      if @append
+        "#{@path_prefix}#{id}#{@path_suffix}#{suffix}"
```

Comment:
If `chunk.metadata` is empty, `id` becomes chunk unique id.
So this path is always unique, and never be appended any other chunk contents. Is it intentional?